File: binding/src/network/libqhstspolicy.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qhstspolicy.html
pub const qhstspolicy = struct {

    /// New constructs a new QHstsPolicy object.
    pub fn New() QtC.QHstsPolicy {
        return qtc.QHstsPolicy_new();
    }


    /// New2 constructs a new QHstsPolicy object.
    pub fn New2(expiry: ?*anyopaque, flags: i32, host: []const u8) QtC.QHstsPolicy {
        const host_str = qtc.libqt_string{
    .len = host.len,
    .data = host.ptr,
};
return qtc.QHstsPolicy_new2(@ptrCast(expiry), @intCast(flags), host_str);
    }


    /// New3 constructs a new QHstsPolicy object.
    pub fn New3(rhs: ?*anyopaque) QtC.QHstsPolicy {
        return qtc.QHstsPolicy_new3(@ptrCast(rhs));
    }


    /// New4 constructs a new QHstsPolicy object.
    pub fn New4(expiry: ?*anyopaque, flags: i32, host: []const u8, mode: i32) QtC.QHstsPolicy {
        const host_str = qtc.libqt_string{
    .len = host.len,
    .data = host.ptr,
};
return qtc.QHstsPolicy_new4(@ptrCast(expiry), @intCast(flags), host_str, @intCast(mode));
    }


    pub fn OperatorAssign(self: ?*anyopaque, rhs: ?*anyopaque) void {
        qtc.QHstsPolicy_OperatorAssign(@ptrCast(self), @ptrCast(rhs));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QHstsPolicy_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn SetHost(self: ?*anyopaque, host: []const u8) void {
        const host_str = qtc.libqt_string{
    .len = host.len,
    .data = host.ptr,
};
qtc.QHstsPolicy_SetHost(@ptrCast(self), host_str);
    }

    pub fn Host(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHstsPolicy_Host(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhstspolicy.Host: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetExpiry(self: ?*anyopaque, expiry: ?*anyopaque) void {
        qtc.QHstsPolicy_SetExpiry(@ptrCast(self), @ptrCast(expiry));
    }

    pub fn Expiry(self: ?*anyopaque, ) QtC.QDateTime {
        return qtc.QHstsPolicy_Expiry(@ptrCast(self));
    }

    pub fn SetIncludesSubDomains(self: ?*anyopaque, include: bool) void {
        qtc.QHstsPolicy_SetIncludesSubDomains(@ptrCast(self), include);
    }

    pub fn IncludesSubDomains(self: ?*anyopaque, ) bool {
        return qtc.QHstsPolicy_IncludesSubDomains(@ptrCast(self));
    }

    pub fn IsExpired(self: ?*anyopaque, ) bool {
        return qtc.QHstsPolicy_IsExpired(@ptrCast(self));
    }

    pub fn SetHost2(self: ?*anyopaque, host: []const u8, mode: i32) void {
        const host_str = qtc.libqt_string{
    .len = host.len,
    .data = host.ptr,
};
qtc.QHstsPolicy_SetHost2(@ptrCast(self), host_str, @intCast(mode));
    }

    pub fn Host1(self: ?*anyopaque, options: u32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QHstsPolicy_Host1(@ptrCast(self), @intCast(options));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qhstspolicy.Host1: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QHstsPolicy_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/hstspolicy.html#types
pub const enums = struct {
    pub const PolicyFlag = enum {
        pub const IncludeSubDomains: i32 = 1;
    };

};

File: binding/src/network/libqhstspolicy.cpp
#include <QDateTime>
#include <QHstsPolicy>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qhstspolicy.h>
#include "libqhstspolicy.h"
#include "libqhstspolicy.hxx"

QHstsPolicy* QHstsPolicy_new() {
	 return new QHstsPolicy();
}

QHstsPolicy* QHstsPolicy_new2(const QDateTime* expiry, int flags, const libqt_string host) {
	 return new QHstsPolicy(*expiry, static_cast<QFlags<QHstsPolicy::PolicyFlag>>(flags), QString::fromUtf8(host.data, host.len));
}

QHstsPolicy* QHstsPolicy_new3(const QHstsPolicy* rhs) {
	 return new QHstsPolicy(*rhs);
}

QHstsPolicy* QHstsPolicy_new4(const QDateTime* expiry, int flags, const libqt_string host, int mode) {
	 return new QHstsPolicy(*expiry, static_cast<QFlags<QHstsPolicy::PolicyFlag>>(flags), QString::fromUtf8(host.data, host.len), static_cast<QUrl::ParsingMode>(mode));
}

void QHstsPolicy_OperatorAssign(QHstsPolicy* self, const QHstsPolicy* rhs) {
	self->operator=(*rhs);
}

void QHstsPolicy_Swap(QHstsPolicy* self, QHstsPolicy* other) {
	self->swap(*other);
}

void QHstsPolicy_SetHost(QHstsPolicy* self, const libqt_string host) {
	self->setHost(QString::fromUtf8(host.data, host.len));
}

libqt_string QHstsPolicy_Host(const QHstsPolicy* self) {
	QString _ret = self->host();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QHstsPolicy_SetExpiry(QHstsPolicy* self, const QDateTime* expiry) {
	self->setExpiry(*expiry);
}

QDateTime* QHstsPolicy_Expiry(const QHstsPolicy* self) {
	return new QDateTime(self->expiry());
}

void QHstsPolicy_SetIncludesSubDomains(QHstsPolicy* self, bool include) {
	self->setIncludesSubDomains(include);
}

bool QHstsPolicy_IncludesSubDomains(const QHstsPolicy* self) {
	return self->includesSubDomains();
}

bool QHstsPolicy_IsExpired(const QHstsPolicy* self) {
	return self->isExpired();
}

void QHstsPolicy_SetHost2(QHstsPolicy* self, const libqt_string host, int mode) {
	self->setHost(QString::fromUtf8(host.data, host.len), static_cast<QUrl::ParsingMode>(mode));
}

libqt_string QHstsPolicy_Host1(const QHstsPolicy* self, unsigned int options) {
	QString _ret = self->host(static_cast<QFlags<QUrl::ComponentFormattingOption>>(options));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QHstsPolicy_Delete(QHstsPolicy* self) {
    delete self;
}

